channeldb/graph.go

```diff
@@ -1697,6 +1699,14 @@ func (c *ChannelGraph) FilterChannelRange(startHeight, endHeight uint32) ([]uint
 			// we'll convert it into an integer and add it to our
 			// returned set.
 			cid = byteOrder.Uint64(k)
+			if cid == 0 {
```

Comment:
thinking about this again, perhaps the correct fix is to disallow inserting `cid = 0` to begin with?

if these are the result of restored channels, and multiple of them have a zero short channel id, are they overwriting each other?

i think if we disallow zero short channel ids, then that will point to the real source of the issue which is that the data being inserted doesn't seem to be fully initialized.

Reply:
So in the past, a zero short channel ID will only be set if we make an SCB when the channel is still unconfirmed (as of lnd 0.9). This was fixed in 0.8.2 in this PR: https://github.com/lightningnetwork/lnd/pull/3719

Reply:
If this is still happening, then there might be some other issue here. However we could still lead forward an include this change if it'll restore some more stability to the itests, and possible peer connections in the network itself.

Reply:
Looks like this is still happening because [we set the block height back to zero]( https://github.com/lightningnetwork/lnd/pull/3719/files#diff-7f4a277cb277edad61783a1df7301b19R198), and the txindex and output index are also zero.

I think the immediate problem, `"current sid: 0:0:0 isn't greater than last sid 0:0:0".`, can be addressed by removing that line of code, while also adding a check for zero-value short channel ids during insertion to prevent any future instances of this case. The main question then is whether or not the disconnect will still occur because the claimed short channel id won't pass validation when the remote node fetches the utxo.

There are a few approaches we can take here:
 1. Wait for restored, pending channels to confirm before inserting the channel shells. If the channel still hasn't confirmed the coins can still be swept via `--sweepall` since the wallet shouldn't know about the conflicting transaction.
 1. Similar to what is already implemented, filter out coinbase channels (`tx_index == 0`) from `FilterChannelRange`, so that we only send info about non-restored channels.
 1. Forbid gossip syncs entirely after restoring channels. This would be pretty easy when restoring on startup, a little more difficult if it's done post-facto via rpc, but not terrible.

1 is probably the best solution since it more closely mimcs the flow when the channel was initially created. However, 2 is the most direct and IMO addresses the issue well enough for the time being. The main caveat that it does start to leak hacks made at the SCB level into other parts of the codebase, manifesting as yet another hack. 3 is probably worth investigating in its own right, but also seems more brittle and invasive compared 1 and 2 on its own.

I would tentatively say we should pursue 2 in this PR, and we can follow up with 1 at a later date—would like to hear some other thoughts on this though.

There's also then the separate (follow up) issue of how to handle multiple pending channels at the same broadcast height, which is currently unsupported. If we choose not to go with option 1,  we could use a simple counter to insert shells at distinct output-indexes on our phantom-coinbase-funding transaction, and the fixes in 2 or 3 should still work.

> However we could still lead forward an include this change if it'll restore some more stability to the itests 

My personal preference would be to fix the root cause as much as possible in one go rather than the symptoms, since elsewhere in the codebase we critically rely on the assumption that short channel ids are non-zero. Considering also that the bug was unaddressed for ~4 months while the itests _were_ failing, I'd rather not remove our primary incentive to investigate a proper solution.

> possible peer connections in the network itself.

Fortunately idt this is that prevalent on the network today, can't be that many people restoring from SCB with pending channels.

Reply:
I agree with @cfromknecht's assessment for the most part (but see discussion below).

Alternative 3 seems pretty hard as you note because you could have done something to your node (opened a new channel) before importing the SCB. This seems very much not recommended but could happen in which case you'd then need to decide to either disable operations after importing the backup or segregate allowed operations based on the status of the channel.

Is there any discussion on what to do to a node after restoring an SCB and completing the DLPs?  Is it recommended to continue operations with the same node or to create a new one? Or node operators are responsible for evaluating this?

> My personal preference would be to fix the root cause as much as possible in one go rather than the symptoms

While I agree in general to this principle, I still think a guard such as the one suggested in this PR should be added (maybe even going back to the previous version where the check is done higher up in the stack, in the gossiper vs the channeldb) to ensure our internal representation of restored/unconfirmed channels don't end up regressing this bug again. 

Ideally we should clarify in BOLT-0007 whether 0-sid channels are valid as members of `encoded_short_channel_id` and if receivers can rely on senders never sending them (and fail their connection if they ever do) **or alternatively**, if receivers should be expected to handle this case appropriately (and then the correct place to fix this particular issue in lnd wouldn't even be here but rather in lnwire, during parsing of the encoded sids).

Reply:
Btw, I'm also ok with having this guard only added at the end of a release cycle so that we don't prematurely hide the issue ("fix the symptoms vs the root cause" as you wrote).

Reply:
Another option I played around with last week was not adding these restored channels to the graph in the first place since the channels won't be usable. @Roasbeef mentioned that when initially working on this, there were some assumptions throughout the codebase that relied on the channel existing in the graph, which is why it's there now. I tried removing the relevant code that adds the restored channels to the graph and ran the integration test suite locally. They all passed, so things might've changed since then. Granted this may not be enough reassurance to consider the change, it may be worth looking into.

Reply:
> Is there any discussion on what to do to a node after restoring an SCB and completing the DLPs? Is it recommended to continue operations with the same node or to create a new one? Or node operators are responsible for evaluating this?

Once the funds are swept it is possible to use the same seed, though typically a new node is recommended to avoid having the left over channel shells. Some people send their funds entirely to a new seed since it provides an opportunity to have a fresh start so to speak. @Roasbeef can probably comment more if i've left out any details here.

> I still think a guard such as the one suggested in this PR should be added

I think we are in agreement here, assuming these channel shells are still inserted into the graph, we should filter out any restored, pending sids. IIUC the main question is whether they are all-zeros (like today) or whether they are these phantom coinbase outputs (`tx_index = 0`) that include the broadcast height as their block height to make them non-zero. The latter however is filtering a super super set that includes the all-zeros sid.

If we were to go with the latter approach, the check for preventing the all-zeroes sid becomes orthogonal to the problem at hand. Its primary purpose would be to prevent similar assumptions from being broken in other subsystems, and in doing so we'd be forced to insert the pending restored channels under some other non-zero sids _and_ keep some level of filtering in the gossiper. I would be okay in moving that check higher up the stack as well 👍

As @wpaulino notes though, it seems possible to avoid inserting these into the graph entirely which would solve many of these headaches in one fell swoop. That may require more investigation though, it could be that the integration tests aren't thorough enough to  show whether or not inserting channel shells is necessary.

> Ideally we should clarify in BOLT-0007 whether 0-sid channels are valid as members of encoded_short_channel_id and if receivers can rely on senders never sending them (and fail their connection if they ever do) if receivers should be expected to handle this case appropriately (and then the correct place to fix this particular issue in lnd wouldn't even be here but rather in lnwire, during parsing of the encoded sids).

I think you are correct here, it is somewhat ambiguous in the spec. One the one hand, we know that the all-zeros SID cannot be a valid channel, but it is ultimately as you say a question of whether that should be a wire parsing error or a validation error.

My personal interpretation would be that failing to parse the all-zero sid is actually an LND bug, rather than a spec issue. It isn't technically possible for the first element in a stream to be unsorted, so I think there is an alternative/additional fix here in that [this check](https://github.com/lightningnetwork/lnd/blob/master/lnwire/query_short_chan_ids.go#L188) should only apply when `i > 0`. Then any dependent messages will fail during validation instead of causing the connection to be broken.

Thanks again for your thorough replies 😀

Reply:
> I personally also would like to see @wpaulino's suggestion implemented as that removes most causes for headaches.

See https://github.com/lightningnetwork/lnd/pull/4370.